### ISSUE TITLE
Inconsistent mAP50-95 between training validation and model.val() for COCO segmentation

### ISSUE THREAD

**Issue by qiyuliaogh** [ultralytics/ultralytics#22149]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Val

### Bug

I'm training a YOLO segmentation model on COCO dataset and getting significantly different mAP50-95 scores:

During training (end of epoch validation): 47.3%

<img width="954" height="105" alt="Image" src="https://github.com/user-attachments/assets/6e7014cc-472d-4eff-b841-8af8ede33ce3" />

Using model.val() with the saved weights: 44.7% (-2.6%)

<img width="616" height="70" alt="Image" src="https://github.com/user-attachments/assets/6846d318-f28e-4aea-9b53-11d8f6f26b87" />

Using fast-coco-eval library: 43.2% (-4.1% from training)

<img width="815" height="85" alt="Image" src="https://github.com/user-attachments/assets/4a018017-3f6e-4446-87e5-c0615793f20c" />

All three methods are evaluating the same model weights on the same COCO validation set, but the results vary by over 4%. This makes it unclear which metric to report when comparing with other models.

What causes this validation discrepancy and which result should be used for fair comparison with other published models? Is this expected behavior or a potential bug?

### Environment

I've been trying on different environments

### Minimal Reproducible Example

# Training shows mAP50-95 = 47.3% during validation
model = YOLO('yolo12x-seg.pt')
model.train(data='coco.yaml', epochs=100, ...)

# But model.val() shows mAP50-95 = 44.7% and 43.2%
model = YOLO('runs/segment/train/weights/best.pt')
model.val(data='coco.yaml')

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @qiyuliaogh, thank you for your interest in Ultralytics 🚀! This is an automated response — an Ultralytics engineer will review and assist you shortly.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

Since you’re seeing a discrepancy in validation metrics, please include:
-Exact Ultralytics, Python, and PyTorch versions, plus CUDA/CUDNN details
-Full training and validation commands, including all args (imgsz, batch, conf, iou, device, seed, workers, augment, rect, half, plots, save_dir, etc.)
-The specific weights file path you evaluated and corresponding training run directory
-Full training logs (end-of-epoch val lines) and the exact `model.val()` output
-Any custom data.yaml modifications and COCO split details used
-If applicable, the exact code snippet for any third-party evaluation you ran
-A small MRE (few images + script) that reproduces the difference

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a Python>=3.8 environment with PyTorch>=1.8 to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including CUDA/CUDNN, Python and PyTorch preinstalled):

- Notebooks with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- Google Cloud Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- Amazon Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- Docker Image. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
Ultralytics mAP would be different from COCOEval mAP. You can read this

https://github.com/ultralytics/ultralytics/issues/17270#issuecomment-2503639995

**Comment by fedyhajali** [comment]
> Ultralytics mAP would be different from COCOEval mAP. You can read this
> 
> [#17270 (comment)](https://github.com/ultralytics/ultralytics/issues/17270#issuecomment-2503639995)


@Y-T-G Is the inconsistency between validation during training and model.val() at the end also caused by the issue mentioned in your linked comment? 

Or are you referring only to the difference between model.val() and validation with COCOEval?

**Comment by Y-T-G** [comment]
No. The mAP at the end of training is calculated on FP16 model which can sometimes be different
